### ISSUE TITLE
Float16 mixed precision training

### ISSUE THREAD

**Issue by HLneoh** [tensorflow/tensorflow#61771]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

2.13.0

### Custom code

Yes

### OS platform and distribution

Ubuntu 20.04.5 LTS

### Python version

Python 3.10.9

### GPU model and memory

A6000

### Current behavior?

Enabled float16 training by setting the mixed precision policy, but why I still need to manually cast the y tensors to float16 before calculating the loss?

Error when no manual cast the tensor:
![image](https://github.com/tensorflow/tensorflow/assets/25906607/46c5efcb-f2c2-4b7a-9483-46656009071b)

### Standalone code to reproduce the issue

```shell
Confidential.
```

**Comment by sushreebarsa** [comment]
@HLneoh 
In order to expedite the trouble-shooting process, please provide a code snippet to reproduce the issue reported here. Thank you!
